### PR TITLE
DYN-7341: Pm redesign search tab cherrypick

### PR DESCRIPTION
**Pull request by dnenov**
### Purpose

Follow [DYN-7341](https://jira.autodesk.com/browse/DYN-7341). Related tasks:
- https://jira.autodesk.com/browse/DYN-7342 

A redesign of the PM search & install functionality. 
- package item row redesign
- package item details redesign
- introduced version compatibility (against Dynamo and Host)
- release notes
- incompatible and unknown compatibility installation warning prompt

### Current State

iimFy4tMRG

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [x] This PR modifies some build requirements and the readme is updated
- [x] This PR contains no files larger than 50 MB

### Release Notes

- cherry picking WIP changes
- contains Greg update to 3.0.2.6334
- refactoring some of the view model code to simplify and remove unnecessary complexity
- added compatibility tests
- added test for latest compatible version
- will set the detail version expander corresponding to the currently selected package version to expanded upon opening the package details view
- changed invoking of the scroll up to happen after the view has finished generating to make sure it scrolls all the way to the top
- added tooltips and related resources
- added fallback message to compatibility section inside detail items if no compatibility is found
- properly centers the messagebox in Dynamo/Package Manager window

### Reviewers

@QilongTang 
@zeusongit

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 1]
```diff
@@ -749,11 +748,28 @@ private bool WarnAboutDuplicatePackageConflicts(PackageVersion package,
 
         internal async void ExecutePackageDownload(string name, PackageVersion package, string installPath)
         {
-            string msg = String.IsNullOrEmpty(installPath) ?
+            string msg;
+            MessageBoxResult result;
+
+            var compatible = PackageManagerSearchElement.CalculateCompatibility(package.compatibility_matrix); 
+            if (compatible == null || compatible == false)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:755]: For nullable bool, looks like we can just use `!= true`, referring https://stackoverflow.com/questions/2673918/best-way-to-check-for-nullable-bool-in-a-condition-expression-if#:~:text=If%20you%20want%20to%20treat%20a%20null%20as%20false,%20then

Comment (dnenov) [reply src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:755]: Oh, of course! Elegant.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs [hunk 2]
```diff
@@ -806,7 +822,7 @@ internal async void ExecutePackageDownload(string name, PackageVersion package,
 
                 // local package that conflicts with package itself.
                 Package duplicatePackage = null;
-
+
```
Comment (zeusongit) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:825]: spaces

File: src/DynamoPackages/PackageManagerSearchElement.cs [hunk 3]
```diff
@@ -135,6 +189,41 @@ internal set
...
+                {"2021.1", "2.6.1"}, {"2022", "2.10.1"}, {"2022.1", "2.12.0"}, {"2023", "2.13.1"},
+                {"2023.1", "2.16.1"}, {"2023.1.3", "2.16.2"}, {"2024", "2.17.0"}, {"2024.1", "2.18.1"},
+                {"2024.2", "2.19.3"}, {"2025", "3.0.3"}, {"2025.1", "3.0.3"}, {"2025.2", "3.2.1"}
+            }},
+            { "Civil3D", new Dictionary<string, string> {
+                {"2020", "2.1.1"}, {"2020.1", "2.2.0"}, {"2020.2", "2.4.1"}, {"2021", "2.5.2"},
+                {"2022", "2.10.1"}, {"2023", "2.13.1"}, {"2024", "2.17.1"}, {"2024.1", "2.18.1"},
+                {"2024.2", "2.18.1"}, {"2024.3", "2.19"}, {"2025", "3.0.3"}, {"2025.1", "3.2.2"}
+            }}
+        };
```
Comment (zeusongit) [inline src/DynamoPackages/PackageManagerSearchElement.cs:226]: We can use the latest Greg nuget with the new API now available for host mappings.https://www.nuget.org/packages/Greg/3.0.2.6533

**Comment by zeusongit** [comment]
In the GIF when installing a package with unknown compatibility, the message shown is for incompatibility.
Is this expected?
Screenshot 2024-10-15 at 2 44 22 PM

**Comment by reddyashish** [comment]
Also @dnenov there is one conflict, resolve it.

File: src/DynamoCore/DynamoCore.csproj [hunk 4]
```diff
@@ -30,6 +30,8 @@
     <PackageReference Include="System.Drawing.Common" Version="6.0.0" />
   </ItemGroup>
   <ItemGroup Label="Common dependencies">
+    <PackageReference Include="Autodesk.IDSDK" Version="1.1.8" />
```
Comment (reddyashish) [inline src/DynamoCore/DynamoCore.csproj:33]: This IDSDK version was updated recently to 1.2.4. Sync with latest master before merging.

File: src/PackageDetailsViewExtension/PackageDetailItem.cs [hunk 5]
```diff
@@ -187,7 +265,7 @@ public bool IsEnabledForInstall
         /// <param name="packageVersion"></param>
         /// <param name="canInstall"></param>
         /// <param name="isEnabledForInstall">True, if package is not already downloaded, is not deprecated, and package loading is allowed.</param>
-        public PackageDetailItem(string packageName, PackageVersion packageVersion, bool canInstall, bool isEnabledForInstall = true)
+        public PackageDetailItem(List<VersionInformation> versionDetails, string packageName, PackageVersion packageVersion, bool canInstall, bool isEnabledForInstall = true)
```
Comment (reddyashish) [inline src/PackageDetailsViewExtension/PackageDetailItem.cs:268]: API break here, create a overloaded constructor if needed.

File: src/PackageDetailsViewExtension/PackageDetailItem.cs [hunk 6]
```diff
@@ -1,13 +1,20 @@
...
 using Dynamo.PackageManager;
 using Dynamo.PythonServices;
 using Greg.Responses;
-using PythonNodeModels;
 
 namespace Dynamo.PackageDetails
 {
+    public class FlattenedCompatibility
+    {
+        public string CompatibilityName { get; set; }
```
Comment (reddyashish) [inline src/PackageDetailsViewExtension/PackageDetailItem.cs:14]: Add summary to these public methods

**Comment by dnenov** [comment]
This is taken care of, I will leave the gif for posterity, as currently I have nothing to demonstrate the prompt with otherwise :/ But, just so it is clear, only `incompatible` packages will trigger this prompt!

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs [hunk 7]
```diff
@@ -83,14 +117,20 @@ public PackageManagerSearchElementViewModel(PackageManagerSearchElement element,
             CanInstall = install;
             IsEnabledForInstall = isEnabledForInstall;
 
-            this.SelectedVersion = this.SearchElementModel.LatestVersion;
+            // Attempts to show the latest compatible version. If no compatible, will return the latest instead.
+            //this.SelectedVersion = this.SearchElementModel.LatestVersion;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs:121]: Is this line still needed?

**Comment by dnenov** [comment]
`DynamoCoreWpfTests.NodeAutoCompleteSearchTests.NodeSuggestions_CanAutoCompleteOnCustomNodesOutPort_WithSpaceInPortName` this passes locally, not sure why it's failing. 
image

File: src/PackageDetailsViewExtension/PackageDetailItem.cs [hunk 8]
```diff
@@ -237,5 +447,13 @@ private void DetectDependencies()
             PythonVersion = pythonEngineVersions.Count > 0 ? string.Join(", ", pythonEngineVersions) : Dynamo.Properties.Resources.NoneString;
             Hosts = hostDependencies.Count > 0 ? string.Join(", ", hostDependencies) : Dynamo.Properties.Resources.NoneString;
         }
+
+        private static string CapitalizeFirstLetter(string word)
```
Comment (QilongTang) [inline src/PackageDetailsViewExtension/PackageDetailItem.cs:451]: We can maybe move this to utilityclass later

File: src/PackageDetailsViewExtension/PackageDetailItem.cs [hunk 9]
```diff
@@ -209,6 +331,94 @@ public PackageDetailItem(string packageName, PackageVersion packageVersion, bool
...
+                return string.Empty;
+            }
+        }
+
+        /// <summary>
+        /// Try parse a date string in a preferred format
+        /// </summary>
+        /// <param name="dateString"></param>
+        /// <returns></returns>
+        private string GetFormattedDate(string dateString)
```
Comment (QilongTang) [inline src/PackageDetailsViewExtension/PackageDetailItem.cs:396]: same comment here

**Pull request merged**